Modules/posixmodule.c

```diff
@@ -5490,12 +5456,103 @@ os__path_normpath_impl(PyObject *module, PyObject *path)
         return NULL;
     }
     Py_ssize_t norm_len;
-    wchar_t *norm_path = _Py_normpath_and_size(buffer, len, &norm_len);
+    wchar_t *norm_path = _Py_normpath_and_size(buffer, len, 0, &norm_len);
     PyObject *result = PyUnicode_FromWideChar(norm_path, norm_len);
     PyMem_Free(buffer);
     return result;
 }
 
+/*[clinic input]
+os._path_abspath
+
+    path: unicode
+    /
+
+Make path absolute.
+[clinic start generated code]*/
+
+static PyObject *
+os__path_abspath_impl(PyObject *module, PyObject *path)
+/*[clinic end generated code: output=b58956d662b60be0 input=577ecb3473d22113]*/
+{
+    Py_ssize_t path_len, abs_len;
+    wchar_t *abs, *abs_buf = NULL, *cwd_buf = NULL;
+    PyObject *result = NULL;
+
+    wchar_t *path_buf = PyUnicode_AsWideCharString(path, &path_len);
+    if (!path_buf) {
+        goto exit;
+    }
+
+#ifdef MS_WINDOWS
```

Comment:
The call needs to fail with `ValueError` on Windows if the path has embedded null characters.

```c
    if (wcslen(path_buf) != path_len) {
        PyErr_Format(PyExc_ValueError, "_path_abspath: embedded null character in path");
        goto exit;
    }
```

The exception is handled by calling `_abspath_fallback()`.

Reply:
Note that `_abspath_fallback()` misbehaves for drive-relative paths such as "Z:spam/eggs". It fails to resolve the working directory on the drive, unless it's the same drive as that of the current working directory. Getting the working directory on each drive ("A:" to "Z:") requires WinAPI `GetFullPathNameW()`. That could be a reason to keep `nt._getfullpathname()` around. For example:

```python
def _getcwd_drive_fallback(drive):
    drive = os.fspath(drive)
    if isinstance(path, bytes):
        return drive + b'\\'
    else:
        return drive + '\\'

try:
    from nt import _getfullpathname
except ImportError:
    _getcwd_drive = _getcwd_drive_fallback
else:
    def _getcwd_drive(drive):
        try:
            return _getfullpathname(drive)
        except (OSError, ValueError):
            return _getcwd_drive_fallback(drive)
```

Reply:
I think that's premature optimization: calling `_Py_normpath_and_size` on a 2 character path shouldn't take very long.
And it's still handled correctly by `nt._path_abspath()`.